pkg/cache/v3/linear.go

```diff
@@ -21,46 +21,73 @@ import (
 	"strconv"
 	"strings"
 	"sync"
-	"sync/atomic"
 
 	"github.com/envoyproxy/go-control-plane/pkg/cache/types"
 	"github.com/envoyproxy/go-control-plane/pkg/log"
 )
 
+// cachedResource is used to track resources added by the user in the cache.
+// It contains the resource itself and its associated version (currently in two different modes).
+// TODO(valerian-roche): store serialized resource if already marshaled to compute the version.
```

Comment:
OK, a bit more on this: it looks like in our deployment, most of the work the control plane is doing is serializing protobufs. I was going to suggest not using sha256 since that seems inneficient compared to e.g. xxhash or murmur, but this is irrelevant given our current profiles: https://ddstaging.datadoghq.com/profiling/search?my_code=disabled&viz=flame_graph&start=1707393371259&end=1707396971259&paused=false, which represents ~20% of our profiles.

Reply:
Moved the comment to `computeResourceStableVersion`
On the issue at the core of it, there are two aspects:
 - we compute the version of all resources on updates. This is quite expensive and a lot of them are never returned when using linear cache in a model where all potential resources are stored but only a few actually watched
 - otoh, when using snapshot cache or using linear cache tracking actually used keys, we are wasting a lot of CPU as this is usually the largest footprint of the cache. This is even worse with lds/cds where unchanged resources are marshaled each time

The main potential gains I see here are:
 - store the serialized version once computed. This will be simple once simple cache is aligned. Imo we should even drop the unserialized resource as we actually don't use it, and the go representation is likely to be significantly larger in memory. This will also enable a mode when the user would provide opaque resources to the cache
 - use vtprotobuf. The envoy protobufs in the repo include support now, and I'd expect very significant gains when enabled from the nbs provided by istio in https://github.com/envoyproxy/go-control-plane/issues/824. Currently vtprotobuf does not provide access to its (already implemented) deterministic mode, but it has been requested by [istio maintainers](https://github.com/planetscale/vtprotobuf/issues/82)